src/store/fileUploadStore.js

```diff
@@ -371,7 +367,35 @@ const actions = {
 				dispatch('markTemporaryMessageAsFailed', { message: temporaryMessage, reason: 'failed-share' })
 				console.error('An error happened when trying to share your file: ', error)
 			}
-		}))
+		}
+
+		const client = getDavClient()
+		const userRoot = '/files/' + getters.getUserId()
+
+		const uploads = getters.getUploadsArray(uploadId)
+		// Check for duplicate names in the uploads array
+		if (findDuplicateUploadNames(uploads)) {
+			// Search for uniquePath and upload files one by one to avoid 423 (Locked)
+			for (const upload of uploads) {
+				await performUpload(upload)
+			}
+		} else {
+			// All original names are unique, upload files in parallel
+			await Promise.all(uploads.map(upload => performUpload(upload)))
+		}
+
+		const shares = getters.getShareableFiles(uploadId)
+		// Check if caption message for share was provided
+		if (caption) {
+			// Share files one by one to keep their order and put caption on the last message
+			for (const share of shares) {
+				await performShare(share)
+			}
```

Comment:
What if you async share all files except the last one in the array and then share that last file ?

Reply:
Good point. Then we could extract last file for share flow and do the same for duplicates in upload flow

Reply:
For duplicate upload, it won't be necessary. The duplication could be in any index ( or indices).

Reply:
You're not wrong) But if we have 10 unique files and 2 with duplicates, we may upload 10 in parallel, and only 2 consequently.

Reply:
Ah I see, you are also not wrong.